subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaBasePlugin.java

```diff
@@ -195,6 +195,12 @@ public Object call() throws Exception {
                 return sourceSet.getCompileClasspath();
             }
         });
+        new DslObject(compileTask.getOptions()).getConventionMapping().map("annotationProcessorPath", new Callable<Object>() {
+            @Override
+            public Object call() throws Exception {
+                return sourceSet.getAnnotationProcessorPath().isEmpty() ? null : sourceSet.getAnnotationProcessorPath();
```

Comment:
I think we shouldn't do an empty check here, but make the detector a little smarter. It can check whether the user passed a simple empty file collection or whether it's looking at a `Configuration`. If it's a `Configuration`, we fall back to the `compileClasspath` and emit a warning if there are processors in there. If it's a plain old file collection, we know that the user explicitly wanted us to ignore the `compileClasspath`. The warning could then instruct users to explicitly set the processor path to empty if they want no processing and tell them that this will be the default behavior in 5.0.